iocore/eventsystem/UnixEThread.cc

```diff
@@ -168,128 +250,12 @@ EThread::execute()
-          process_event(e, e->callback_event);
-        } else if (e->timeout_at > 0) { // INTERVAL
-          EventQueue.enqueue(e, cur_time);
-        } else { // NEGATIVE
-          Event *p = nullptr;
-          Event *a = NegativeQueue.head;
-          while (a && a->timeout_at > e->timeout_at) {
-            p = a;
-            a = a->link.next;
-          }
-          if (!a) {
-            NegativeQueue.enqueue(e);
-          } else {
-            NegativeQueue.insert(e, p);
-          }
-        }
-      }
-      bool done_one;
-      do {
-        done_one = false;
-        // execute all the eligible internal events
-        EventQueue.check_ready(cur_time, this);
-        while ((e = EventQueue.dequeue_ready(cur_time))) {
-          ink_assert(e);
-          ink_assert(e->timeout_at > 0);
-          if (e->cancelled) {
-            free_event(e);
-          } else {
-            done_one = true;
-            process_event(e, e->callback_event);
-          }
-        }
-      } while (done_one);
-      // execute any negative (poll) events
-      if (NegativeQueue.head) {
-        if (n_ethreads_to_be_signalled) {
-          flush_signals(this);
-        }
-        // dequeue all the external events and put them in a local
-        // queue. If there are no external events available, don't
-        // do a cond_timedwait.
-        if (!INK_ATOMICLIST_EMPTY(EventQueueExternal.al)) {
-          EventQueueExternal.dequeue_timed(cur_time, next_time, false);
-        }
-        while ((e = EventQueueExternal.dequeue_local())) {
-          if (!e->timeout_at) {
-            process_event(e, e->callback_event);
-          } else {
-            if (e->cancelled) {
-              free_event(e);
-            } else {
-              // If its a negative event, it must be a result of
-              // a negative event, which has been turned into a
-              // timed-event (because of a missed lock), executed
-              // before the poll. So, it must
-              // be executed in this round (because you can't have
-              // more than one poll between two executions of a
-              // negative event)
-              if (e->timeout_at < 0) {
-                Event *p = nullptr;
-                Event *a = NegativeQueue.head;
-                while (a && a->timeout_at > e->timeout_at) {
-                  p = a;
-                  a = a->link.next;
-                }
-                if (!a) {
-                  NegativeQueue.enqueue(e);
-                } else {
-                  NegativeQueue.insert(e, p);
-                }
-              } else {
-                EventQueue.enqueue(e, cur_time);
-              }
-            }
-          }
-        }
-        // execute poll events
-        while ((e = NegativeQueue.dequeue())) {
-          process_event(e, EVENT_POLL);
-        }
-        if (!INK_ATOMICLIST_EMPTY(EventQueueExternal.al)) {
-          EventQueueExternal.dequeue_timed(cur_time, next_time, false);
```

Comment:
@SolidWallOfCode The 3rd parameter is false. It means `sleep == false` and `ink_cond_timedwait(&might_have_data, &lock, &ts);` never be called.

```
 ProtectedQueue::dequeue_timed(ink_hrtime cur_time, ink_hrtime timeout, bool sleep)
 {
    (void)cur_time;
    Event *e;		
    if (sleep) {
      ink_mutex_acquire(&lock);
      if (INK_ATOMICLIST_EMPTY(al)) {
        timespec ts = ink_hrtime_to_timespec(timeout);
        ink_cond_timedwait(&might_have_data, &lock, &ts);
      }
      ink_mutex_release(&lock);
    }
```